mailserver/registry/verifier_test.go

```diff
@@ -0,0 +1,115 @@
+package registry
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind/backends"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/stretchr/testify/suite"
+)
+
+type VerifierTestSuite struct {
+	suite.Suite
+	backend         bind.ContractBackend
+	privKey         *ecdsa.PrivateKey
+	from            common.Address
+	contractAddress common.Address
+	registry        *Registry
+	verifier        *RegistryVerifier
+}
+
+func (s *VerifierTestSuite) SetupTest() {
+	s.setupAccount()
+	s.setupBackendAndContract()
+
+	var err error
+	s.verifier, err = NewVerifier(s.backend, s.contractAddress)
+	s.Require().NoError(err)
+	s.verifier.pendingOpt = true
+}
+
+func (s *VerifierTestSuite) setupBackendAndContract() {
+	var err error
+
+	auth := bind.NewKeyedTransactor(s.privKey)
+	alloc := make(core.GenesisAlloc)
+	alloc[auth.From] = core.GenesisAccount{Balance: big.NewInt(133700000)}
+	s.backend = backends.NewSimulatedBackend(alloc)
+
+	auth = s.newKeyedTransactor()
+
+	s.contractAddress, _, s.registry, err = DeployRegistry(auth, s.backend)
+	s.Require().NoError(err)
+}
+
+func (s *VerifierTestSuite) setupAccount() {
+	var err error
+
+	s.privKey, err = crypto.GenerateKey()
+	s.Require().NoError(err)
+
+	pubKey := s.privKey.Public()
+	publicKeyECDSA, ok := pubKey.(*ecdsa.PublicKey)
+	s.Require().True(ok)
+
+	s.from = crypto.PubkeyToAddress(*publicKeyECDSA)
+}
+
+func (s *VerifierTestSuite) add(nodeID discover.NodeID) {
+	auth := s.newKeyedTransactor()
+	_, err := s.registry.Add(auth, nodeID[:])
+	s.Require().NoError(err)
+}
+
+func (s *VerifierTestSuite) newKeyedTransactor() *bind.TransactOpts {
+	nonce, err := s.backend.PendingNonceAt(context.Background(), s.from)
+	s.Require().NoError(err)
+
+	gasPrice, err := s.backend.SuggestGasPrice(context.Background())
+	s.Require().NoError(err)
+
+	auth := bind.NewKeyedTransactor(s.privKey)
```

Comment:
there is one difference how transactor for ropsten/mainnet needs to be constructed. both ropsten/mainnet expects txs to be signed with chain id (eip155), but simulated backend doesn't. I am using this:

```
func Opts(key *ecdsa.PrivateKey, chain *big.Int, price *big.Int, limit uint64) *bind.TransactOpts {
	addr := crypto.PubkeyToAddress(key.PublicKey)
	var signer types.Signer
	if chain == nil {
		signer = types.HomesteadSigner{}
	} else {
		signer = types.NewEIP155Signer(chain)
	}
	log.Debug("created signer", "chain", chain, "addr", addr, "price", price, "gas limit", limit)
	return &bind.TransactOpts{
		From:     addr,
		GasPrice: price,
		GasLimit: limit,
		Signer: func(_ types.Signer, address common.Address, tx *types.Transaction) (*types.Transaction, error) {
			if address != addr {
				return nil, errors.New("not authorized to sign this account")
			}
			signature, err := crypto.Sign(signer.Hash(tx).Bytes(), key)
			if err != nil {
				return nil, err
			}
			return tx.WithSignature(signer, signature)
		},
	}
}
```

Reply:
thank you @dshulyak . I see that `NewKeyedTransactor` implements something similar, so maybe I can use that one everywhere. Also the `callContract` of the `SimulatedBackend` sets the `GasPrice` automatically to 1 because it's now needed, so I don't even need to specify `GasLimit` and `GasPrice`.